students/DiannaTingg/lessons/lesson07/assignment/parallel.py

```diff
@@ -0,0 +1,192 @@
+    """
+    Prints all documents in a collection.
+    :param collection_name: collection
+    """
+    for doc in collection_name.find():
+        print(doc)
+
+
+def _import_csv(filename):
+    """
+    Returns a list of dictionaries.  One dictionary for each row of data in a csv file.
+    :param filename: csv file
+    :return: list of dictionaries
+    """
+
+    with open(filename, newline="") as csvfile:
+        dict_list = []
+
+        csv_data = csv.reader(csvfile)
+
+        headers = next(csv_data, None)  # Save the first line as the headers
+
+        if headers[0].startswith("ï»¿"):  # Check for weird formatting
+            headers[0] = headers[0][3:]
+
+        for row in csv_data:
+            row_dict = {column: row[index] for index, column in enumerate(headers)}
+
+            dict_list.append(row_dict)
+
+        return dict_list
+
+
+def _add_bulk_data(results, collection, directory_name, filename):
+    """
+    Adds data in bulk to database.
+    :param collection: collection
+    :param directory_name: directory
+    :param filename: csv file
+    :return: records processed (int), initial records (int), final records (int), function run time (float)
+    """
+    file_path = os.path.join(directory_name, filename)
+
+    start_time = time.time()
+    initial_records = collection.count_documents({})
+
+    collection.insert_many(_import_csv(file_path), ordered=False)
+
+    final_records = collection.count_documents({})
+    records_processed = final_records - initial_records
+    run_time = time.time() - start_time
+
+    stats = (records_processed, initial_records, final_records, run_time)
+
+    results[collection.name] = stats
+
+
+def import_data(db, directory_name, products_file, customers_file, rentals_file):
+    """
+    Takes a directory name and three csv files as input.  Creates and populates three collections in MongoDB.
+    :param db: MongoDB
+    :param directory_name: directory name for files.  Use "" for current directory.
+    :param products_file: csv file with product data
+    :param customers_file: csv file with customer data
+    :param rentals_file: csv file with rentals data
+    :return: List of tuples (one for customers and one for products). Each tuple contains:
+             num of records processed, initial record count, final record count, and module run time
+    """
+
+    products = db["products"]
+    customers = db["customers"]
+    rentals = db["rentals"]
+
+    results_dict = {}
+
+    threads = [threading.Thread(target=_add_bulk_data, args=(results_dict, products, directory_name, products_file)),
+               threading.Thread(target=_add_bulk_data, args=(results_dict, customers, directory_name, customers_file)),
+               threading.Thread(target=_add_bulk_data, args=(results_dict, rentals, directory_name, rentals_file))]
+
+    for t in threads:
+        t.start()
+
+    for t in threads:
+        t.join()
+
+    return [results_dict["customers"], results_dict["products"]]
```

Comment:
Using a dictionary is a fine solution. If you are using lots of threads, or threads need to share data, queue or dequeue is the answer.